### PR TITLE
Caffe Plugin

### PR DESCRIPTION
**Pull request by piiswrong**
@HrWangChengdu I helped you created this pull request to track progress and make some comments.

### CODE REVIEW

File: plugin/caffe/caffe_blob.h [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+#include<caffe/blob.hpp>
+#include<mshadow/tensor.h>
+#include<mshadow/tensor_blob.h>
+#include<vector>
+
+namespace mxnet {
+namespace op {
+
+using mshadow::TBlob;
```
Comment (piiswrong) [inline plugin/caffe/caffe_blob.h]: Don't put global using in headers under common namespace. Consider moving them into functions or mxnet/op/caffe namespace

File: plugin/caffe/caffe_blob.h [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+using mshadow::TShape;
+using mshadow::index_t;
+using caffe::Blob;
+// Declare Memory Type for Caffe blob
+namespace caffememtype {
```
Comment (piiswrong) [inline plugin/caffe/caffe_blob.h]: consider rename to simple caffe for consistency

File: plugin/caffe/caffe_blob.h [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+enum caffeMemoryTypes {Data, Grad, Non};
+}  // caffememtype
+
+TShape Vector2TShape(const std::vector<int> &vec_int);
+std::vector<int> TShape2Vector(const TShape &tshape);
+
+// implementation of tensor to blob, called by TensorToBlob
+template<typename Device, typename Dtype>
+void SetDataGradToBlob(caffememtype::caffeMemoryTypes memType,
```
Comment (piiswrong) [inline plugin/caffe/caffe_blob.h]: Consider putting these functions into caffe namespace

File: plugin/caffe/caffe_common.h [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+#ifndef PLUGIN_CAFFE_CAFFE_COMMON_H_
+#define PLUGIN_CAFFE_CAFFE_COMMON_H_
+
+#include <mxnet/operator.h>
+#include<caffe/blob.hpp>
+#include<vector>
+
+namespace mxnet {
+
+using caffe::Blob;
```
Comment (piiswrong) [inline plugin/caffe/caffe_common.h]: same

File: plugin/caffe/caffe_fieldentry.h [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+#include <typeinfo>
+#include <string>
+#include <vector>
+#include <algorithm>
+#include <utility>
+
+namespace dmlc {
+namespace parameter {
+
+using caffe::LayerParameter;
```
Comment (piiswrong) [inline plugin/caffe/caffe_fieldentry.h]: same

File: plugin/caffe/caffe_loss-inl.h [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+          << "Must init CuBLAS handle in stream";
+#endif  // __CUDACC__
+
+    TBlob2CaffeBlob<xpu, Dtype>(caffememtype::Grad, bot_.begin(), in_grad.begin(), param_.in_num);
+    // Pass grad scale to caffe blob
+    top_[0]->set_cpu_diff(&grad_scale_);
+
+    // Set BP flag 
+    for (index_t i = 0; i < param_.in_num; ++i)
+      flags_[i] = req[i] != kNullOp;
```
Comment (piiswrong) [inline plugin/caffe/caffe_loss-inl.h]: caffe doesn't support addto right? should check here.
Also I forgot how caffe handles inplace. Probably need to check too

File: plugin/caffe/caffe_operator-inl.h [hunk 7]
```diff
@@ -0,0 +1,331 @@
...
+  }
+};
+
+
+/**
+ * \brief this is the implementation of caffe operator in caffe.
+ * \tparam xpu the device that the op will be executed on.
+ */
+template<typename xpu, typename Dtype>
+class CaffeOperator : public Operator {
```
Comment (piiswrong) [inline plugin/caffe/caffe_operator-inl.h]: Can caffeLoss and caffeoperator be merged? I think they are all inherited from caffe::Layer?

File: plugin/caffe/caffe_operator-inl.h [hunk 8]
```diff
@@ -0,0 +1,331 @@
...
+    for (index_t i = 0; i < param_.in_num; ++i)
+      flags_[i] = req[i] != kNullOp;
+
+    caffeOp_->Backward(top_, flags_, bot_);
+  }
+
+  void HandleOpReq(mshadow::Stream<xpu>*s, OpReqType req, const TBlob* in_g) {
+    if ((req == kWriteInplace) || (req == kWriteTo))
+      switch (in_g->shape_.ndim()) {
+        case 1: CleanTBlob<1>(s, req, in_g); break;
```
Comment (piiswrong) [inline plugin/caffe/caffe_operator-inl.h]: This is not needed. You can call FlatTo2D if you only want to set it to 0

File: plugin/caffe/caffe_operator_util.cc [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+#include "caffe/layers/cudnn_sigmoid_layer.hpp"
+#include "caffe/layers/cudnn_softmax_layer.hpp"
+#include "caffe/layers/cudnn_tanh_layer.hpp"
+#endif
+
+namespace mxnet {
+namespace op {
+
+
+MXNET_REGISTER_CAFFE_LAYER(InnerProduct, caffe::InnerProductLayer);
```
Comment (piiswrong) [inline plugin/caffe/caffe_operator_util.cc]: Do users have to add their layer here to use it? Can you directly use caffe's registry?

File: python/mxnet/metric.py [hunk 10]
```diff
@@ -307,6 +307,16 @@ def update(self, _, preds):
             self.sum_metric += pred.asnumpy().mean()
         self.num_inst += 1
 
+class Caffe(EvalMetric):
```
Comment (piiswrong) [inline python/mxnet/metric.py]: Consider change to inherit Torch metric

**Comment by piiswrong** [comment]
@tqchen @antinucleon Could you have a look?

File: plugin/caffe/caffe_operator-inl.h (hunk 7 above)
Comment (HrWangChengdu) [inline plugin/caffe/caffe_operator-inl.h]: Yes, they could be merged but I'm afraid then the readability might be a problem.

The major differences between two classes are:
* op has weights to process while loss doesn't
* loss needs to set grad_scale explicitly while op passes out_grad
* arguments of operator-property. op returns "data_i" while loss returns "data & label"

I feel the code logic and class definition would be clearer if written separately. What do you think?

Comment (piiswrong) [inline plugin/caffe/caffe_operator-inl.h]: That also makes sense. Let's leave it as is for now. If you find youself doing too much copy-pasting later consider merge them or share code through inheritance.

File: plugin/caffe/caffe_loss-inl.h (hunk 6 above)
Comment (HrWangChengdu) [inline plugin/caffe/caffe_loss-inl.h]: It depends. Caffe does addto for weight-grad and directly writes to input-grad in its backward. This is reasonable as one might do multiple ff-bp and update net by average gradient.
Thus, if req type for weights gradient are write-to, gradient should be set to zero before doing caffeop backward(). Otherwise, if req is add-to, bakcward() could be called directly.
Also inplace option of caffeop is disabled since it supports various caffe-layers and their conditions differ.

I guess how caffe handles inplace might not matter. Because plugin just passes the dptr_ of tblob to caffe blob and only layer-level interface of caffe is called, I think its inplace implementation won't be called in our code...

File: plugin/caffe/caffe_operator-inl.h (hunk 7 above)
Comment (HrWangChengdu) [inline plugin/caffe/caffe_operator-inl.h]: Sure, no problem.

File: plugin/caffe/caffe_operator_util.cc (hunk 9 above)
Comment (HrWangChengdu) [inline plugin/caffe/caffe_operator_util.cc]: Yes. Users need to add a few lines to register.
For the latter question: one difference between mxnet and caffe is, Mxnet needs to know the exact number of weights before receiving real data (Tblob). While caffe calculates that given exact input data. 
So I made my own registry list in order to set the weight number info.

Comment (piiswrong) [inline plugin/caffe/caffe_operator_util.cc]: The purpose of caffe plugin is to let users migrate from caffe easily. That means they'll be using this to run their own layers, which is not defined here.

It's more difficult for them to add code here and recompile than, say, specify that in a parameter, like is done for torch module.
We can even add a logic in caffe_converter that counts the number of input a layer is having and feed it to the symbol.

tl;dr; I think adding users specify how many input this layer has is better than hard code it here.

Comment (piiswrong) [inline plugin/caffe/caffe_operator_util.cc]: I guess you need both num_inputs and num_weights

File: plugin/caffe/caffe_blob.h (hunk 1 above)
Comment (HrWangChengdu) [inline plugin/caffe/caffe_blob.h]: fixed

File: plugin/caffe/caffe_blob.h (hunk 2 above)
Comment (HrWangChengdu) [inline plugin/caffe/caffe_blob.h]: fixed

File: plugin/caffe/caffe_blob.h (hunk 3 above)
Comment (HrWangChengdu) [inline plugin/caffe/caffe_blob.h]: fixed

File: plugin/caffe/caffe_common.h (hunk 4 above)
Comment (HrWangChengdu) [inline plugin/caffe/caffe_common.h]: fixed

File: plugin/caffe/caffe_fieldentry.h (hunk 5 above)
Comment (HrWangChengdu) [inline plugin/caffe/caffe_fieldentry.h]: fixed

File: plugin/caffe/caffe_operator_util.cc (hunk 9 above)
Comment (HrWangChengdu) [inline plugin/caffe/caffe_operator_util.cc]: I agree that it would be easy to specify those parameters in symbol interface than hard code. Thx!

**Comment by piiswrong** [comment]
@HrWangChengdu Is this currently functional?

Could you give a list of things needed before we can merge in a functional initial version?

**Comment by HrWangChengdu** [comment]
Sorry, I'm in the course due of my graduate school, would be available in 1 or 2 days. Thanks for understanding.
Yes. But only forked dmlc/caffe would work, which add several extra interfaces. In the release version, we could provide a patch for those extra codes.
List of things before we can merge:
- replace by caffe registry
- refine caffe-plugin docs (also interface discussion)
- pull request dmlc/caffe to get patch number
- python metric modification

**Comment by piiswrong** [comment]
What did you add? Is it possible to do it without adding anything to caffe?

**Comment by piiswrong** [comment]
Ok I see. You had to access the private blobs. Looks like simple addition of 3 functions.
In that case let's try PR this back to caffe & generate a patch for users. Should be an easy patch.

**Comment by HrWangChengdu** [comment]
Yes! that's what i mean :-)

**Comment by mli** [comment]
many thanks for the pr, it's very helpful!

I'm not sure caffe will accept the pr quickly. one way is to provide a patch in the readme to tell users how to do the caffe patch manually. 

also is there any test code? such as 
- end-to-end test to check if we can get a reasonable accuray
- unittest to check if same results can be get from mx's op and caffe's op

also i'm curious about the performance. ideally, using caffe's op will have the some performance as mxnet's op. is it true?

**Comment by HrWangChengdu** [comment]
Thank you!

I think we could do the pr to caffe meanwhile provides manually patch document.

We do have. There're mlp & lenet training on mnist by caffe op in example/caffe folder. Both networks achieve high accuracy. And I double checked caffe op by gradient_checker from dmlc/minpy package.

Oh, that would be really interesting. Let me make some performance comparison results this week.

File: plugin/caffe/caffe_operator-inl.h (hunk 8 above)
Comment (HrWangChengdu) [inline plugin/caffe/caffe_operator-inl.h]: Fixed

File: plugin/caffe/caffe_operator_util.cc (hunk 9 above)
Comment (HrWangChengdu) [inline plugin/caffe/caffe_operator_util.cc]: Fixed

File: python/mxnet/metric.py (hunk 10 above)
Comment (HrWangChengdu) [inline python/mxnet/metric.py]: Fixed

**Comment by HrWangChengdu** [comment]
@piiswrong  could you please check python side code, i.e. example/caffe/caffe_net.py?
I guess
 `fc = mx.sym.CaffeOp(data=[data0, data1, data2], prototxt=....)` 
seems better than
 `fc = mx.sym.CaffeOp(data_0=data0, data_1=data1, data_2=data2, prototxt=....)`

File: plugin/caffe/caffe_op-inl.h [hunk 11]
```diff
@@ -16,29 +16,26 @@
...
-  int in_num, w_num, out_num;
+  int data_num, w_num, out_num;
 
-  DMLC_DECLARE_PARAMETER(CaffeOperatorParam) { DMLC_DECLARE_FIELD(prototxt).set_default("layer{}")
+  DMLC_DECLARE_PARAMETER(CaffeOpParam) { DMLC_DECLARE_FIELD(prototxt).set_default("layer{}")
     .describe("Caffe's layer parameter");
-    DMLC_DECLARE_FIELD(in_num).set_range(0, 100).set_default(1)
+    DMLC_DECLARE_FIELD(data_num).set_range(0, 100).set_default(1)
     .describe("Operator input number");
+    DMLC_DECLARE_FIELD(w_num).set_range(0, 100).set_default(0)
```
Comment (piiswrong) [inline plugin/caffe/caffe_op-inl.h]: change to weight_num

File: plugin/caffe/caffe_op-inl.h [hunk 12]
```diff
@@ -16,29 +16,26 @@
...
+struct CaffeOpParam : public dmlc::Parameter<CaffeOpParam> {
   ::caffe::LayerParameter prototxt;
-  int in_num, w_num, out_num;
+  int data_num, w_num, out_num;
 
-  DMLC_DECLARE_PARAMETER(CaffeOperatorParam) { DMLC_DECLARE_FIELD(prototxt).set_default("layer{}")
+  DMLC_DECLARE_PARAMETER(CaffeOpParam) { DMLC_DECLARE_FIELD(prototxt).set_default("layer{}")
     .describe("Caffe's layer parameter");
-    DMLC_DECLARE_FIELD(in_num).set_range(0, 100).set_default(1)
+    DMLC_DECLARE_FIELD(data_num).set_range(0, 100).set_default(1)
```
Comment (piiswrong) [inline plugin/caffe/caffe_op-inl.h]: num_data

File: plugin/caffe/caffe_op-inl.h [hunk 13]
```diff
@@ -16,29 +16,26 @@
...
     .describe("Operator input number");
+    DMLC_DECLARE_FIELD(w_num).set_range(0, 100).set_default(0)
+    .describe("Weight number");
     DMLC_DECLARE_FIELD(out_num).set_range(0, 100).set_default(1)
```
Comment (piiswrong) [inline plugin/caffe/caffe_op-inl.h]: num_out

File: python/mxnet/symbol.py [hunk 14]
```diff
@@ -1048,6 +1048,12 @@ def creator(*args, **kwargs):
         for k, v in kwargs.items():
             if isinstance(v, Symbol):
                 symbol_kwargs[k] = v
+            elif isinstance(v, list):
```
Comment (piiswrong) [inline python/mxnet/symbol.py]: you don't need this. if you have a list of input say [x, y, z,]
you can do caffeop(x, y, z, xxx=yyy, ...) or caffeop(*[x, y, z], xxx=yyy, ...)

**Comment by piiswrong** [comment]
data=[data0, data1, data2] is equivalent to *[data0, data1, data2]

**Comment by piiswrong** [comment]
LGTM. Fix lint and we can merge.

This is a great effort. Thanks so much for doing this.

**Comment by HrWangChengdu** [comment]
Fixed the lint and complement the patch in caffe docs.
Is there anything else to do?
Thank you so much for helping me doing this merge!

**Comment by piiswrong** [comment]
Hopefully the test will pass and I'll merge it

**Comment by piiswrong** [comment]
@HrWangChengdu Because I created this pull request for you, the commit history now traces this to me. You can make a new PR with two commits where the first one removes every file you added in this PR and the second one adds them back in, so that the history will be traced to you.

**Comment by mli** [comment]
you can revert this pr...

**Comment by piiswrong** [comment]
How? I probably don't have the privilige

**Comment by piiswrong** [comment]
@mli I can't do forced push. Can you? Discard the last commit?

**Comment by mli** [comment]
forced push is disabled, which is too dangers. there is a revert button in
the pr page, it basically files another pr to revert all changes

On Tue, Jul 26, 2016 at 10:17 PM, Eric Junyuan Xie <notifications@github.com

**Comment by antinucleon** [comment]
I have created a revert PR

**Comment by HrWangChengdu** [comment]
Should I new a PR after the revert one is done?

**Comment by piiswrong** [comment]
Yes, Please PR gain

**Comment by HrWangChengdu** [comment]
Just create one https://github.com/dmlc/mxnet/pull/2854

**Comment by tqchen** [comment]
It is great that this is merged in. Let us work on making the blogpost to dmlc.ml and put this on the news
